raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -738,17 +739,22 @@ private VoteResponseData buildVoteResponse(
         ListenerName listenerName,
         short apiVersion,
         Errors partitionLevelError,
-        boolean voteGranted
+        boolean voteGranted,
+        boolean preVote
     ) {
+        // We are granting a PreVote, we do not send our leader id if we have one.
+        // This prevents Prospective and Follower states from ping-ponging as described in KIP-996.
+        int leaderId = preVote && voteGranted ? UNKNOWN_LEADER_ID : quorum.leaderIdOrSentinel();
```

Comment:
note: this will be reverted in the next commit. we won't ping-pong between Prospective and Follower for the same Leader if Prospective persists the leaderId